### PR TITLE
Propose different behavior for dealing with the `/s` directory

### PR DESCRIPTION
**Pull request by vtbassmatt**
We talked about a different design for where to checkout sources. This PR reflects that discussion.

### CODE REVIEW

File: design/multi-checkout.md [hunk 1]
```diff
@@ -58,26 +58,30 @@ If it's used again, it's OK to re-fetch the same data.
...
 #### System variables
-`Build.SourcesDirectory` points to the directory where the `self` repo was checked out whether in single- or multi-checkout mode.
-If `self` wasn't checked out, we'll still create an empty `s/` directory and set `Build.SourcesDirectory` to it.
-(This is for back-compat and matches what we do for `checkout: none` today.)
+No change to the contents of system variablies like `Build.SourcesDirectory` or `System.DefaultWorkingDirectory`.
+Those will remain `$(Pipeline.Workspace/s`.
 
-`System.DefaultWorkingDirectory` points to the same place as `Pipeline.Workspace` in multi-checkout mode (which is different than single-checkout mode).
-We believe this preserves the intended semantics of these variables, keeping as many scripts and tasks working as possible.
+**NOTE**: if one or more `checkout`s sets an explicit path outside of `$(Pipeline.Workspace)/s`, it's ambiguous where the "root of source" should be.
```
Comment (vtbassmatt) [inline design/multi-checkout.md]: This is the part I really don't like.

File: design/multi-checkout.md [hunk 2]
```diff
@@ -58,26 +58,30 @@ If it's used again, it's OK to re-fetch the same data.
...
+Two decent heuristics occur to me:
+1. If at least one repo is in a subdirectory of `$(Pipeline.Workspace)/s`, use that as the value for both variables.
+2. Unconditionally set both variables to the same value as `Pipeline.Workspace`.
+
+Of the two, the first is less likely to break tasks.
+The second is easier to explain to pipeline authors.
```
Comment (jpricket) [inline design/multi-checkout.md]: Another alternative, would be to unconditionally set both variables to `$(Pipeline.Workspace)/s` even though it is empty.

File: design/multi-checkout.md (hunk 1 above)
Comment (jpricket) [reply design/multi-checkout.md]: I agree, but without a way for the yaml to specify the "root of source", we just need to make it easy to reason about.

File: design/multi-checkout.md (hunk 2 above)
Comment (vijayma) [reply design/multi-checkout.md]: #(Pipeline.Workspace)/s unconditionally would be my preference as well. The meaning of the $(Build.SourcesDirectory) is the 's' folder.

Comment (vtbassmatt) [reply design/multi-checkout.md]: I think you're both saying that in multicheckout mode we: 
- create the `/s` directory unconditionally
- always set both of those variables to it, regardless of any `path` settings on `checkout` steps

I need to think a bit more about this, but it just might work :grin:

Comment (vtbassmatt) [reply design/multi-checkout.md]: This sounds fine.

File: design/multi-checkout.md [hunk 3]
```diff
@@ -58,26 +58,30 @@ If it's used again, it's OK to re-fetch the same data.
...
-The pipeline workspace directory has three well-known subdirectories:
+The pipeline workspace directory has four well-known subdirectories:
+- `s/` for sources
 - `a/` for artifacts
 - `b/` for generated binaries
 - `testResults/` for test results files
 
-We'll recommend not to use these as checkout paths, but we won't attempt to block them.
-Likewise, a customer could choose to checkout a non-`self` repo into the `s` directory, and we won't block that.
+We'll recommend not to use any of these as checkout paths, but we won't attempt to block them.
```
Comment (leantk) [inline design/multi-checkout.md]: This statement is confusing to me. We are saying use `s/` to checkout correct?

Comment (leantk) [reply design/multi-checkout.md]: Well I'm 6 minutes too late lol

Comment (vtbassmatt) [reply design/multi-checkout.md]: Haha, never too late. The net of it is: in multicheckout, we'll always create a `s/` directory and call it `Build.SourcesDirectory`. By default, code will be checked out into directories under it. The user can override our default by saying `path` on the checkout and put their source wherever they want.

**Pull request merged**